homeassistant/components/synology_dsm/__init__.py

```diff
@@ -181,6 +187,33 @@ def _async_migrator(
     api = SynoApi(hass, entry)
     try:
         await api.async_setup()
+    except (
+        SynologyDSMLogin2SARequiredException,
+        SynologyDSMLoginDisabledAccountException,
+        SynologyDSMLoginInvalidException,
+        SynologyDSMLoginPermissionDeniedException,
+    ) as err:
+        if err.args[0] and isinstance(err.args[0], dict):
+            # pylint: disable=no-member
+            details = err.args[0].get(EXCEPTION_DETAILS, EXCEPTION_UNKNOWN)
+        else:
+            details = EXCEPTION_UNKNOWN
+        _LOGGER.debug(
+            "Reauthentication for DSM '%s' needed - reason: %s",
+            entry.unique_id,
+            details,
+        )
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={
+                    "source": SOURCE_REAUTH,
+                    "data": {**entry.data},
```

Comment:
We normally pass the existing entry data as the `data` parameter instead of in the context, and define a confirm step that is called from the reauth step.